src/Rematch.jl

```diff
@@ -128,42 +128,76 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
             end
         end
     elseif @capture(pattern, T_(subpatterns__))
-        # struct      
-        len = length(subpatterns)
-        named_fields = [pat.args[1] for pat in subpatterns if (pat isa Expr) && pat.head == :kw]
-        named_count = length(named_fields)
-        @assert named_count == length(unique(named_fields)) "Pattern $pattern has duplicate named arguments ($(named_fields))."
-        @assert named_count == 0 || named_count == len "Pattern $pattern mixes named and positional arguments."
-        if named_count == 0
-            # Pattern uses positional arguments to refer to fields e.g. Foo(0, true)
-            expected_fieldcount = gensym("$(T)_expected_fieldcount")
-            actual_fieldcount = gensym("$(T)_actual_fieldcount")
-            push!(asserts, quote
-                if typeof($(esc(T))) <: Function
-                    throw(LoadError("Attempted to match on a function", @__LINE__, AssertionError("Incorrect match usage")))
+        # struct or extractor call
+        if !isempty(string(T)) && islowercase(first(string(T)))
+            # Extractor call.
+            result = gensym("unapply")
+            len = length(subpatterns)
+            # Extractor call.
+            # The function should take one argument and return either `nothing`, if the argument does not match,
+            # or a tuple if it does match.
+            if len == 0
+                # If there are no subpatterns, the result value is just checked for nothingness.
+                quote
+                    $(esc(T))($value) !== nothing
                 end
-                if !(isstructtype(typeof($(esc(T)))) || issabstracttype(typeof($(esc(T)))))
-                    throw(LoadError("Attempted to match on a pattern that is not a struct", @__LINE__, AssertionError("Incorrect match usage")))
+            elseif len == 1
+                # If there is just one subpattern, the result value is matched against it.
+                quote
+                    begin
+                        $result = $(esc(T))($value)
+                        $result !== nothing && $(handle_destruct(result, subpatterns[1], bound, asserts))
+                    end
                 end
-                # This assertion is necessary:
-                # If $expected_fieldcount < $actual_fieldcount, to catch missing fields.
-                # If $expected_fieldcount > $actual_fieldcount, to avoid a BoundsError.
-                $expected_fieldcount = evaluated_fieldcount($(esc(T)))
-                $actual_fieldcount = $(esc(len))
-                if $expected_fieldcount != $actual_fieldcount
-                    error("Pattern field count is $($actual_fieldcount) expected $($expected_fieldcount)")
+            else
+                # If there is more than one subpattern, the result value is matched against a tuple pattern.
```

Comment:
Having looked at the rematch source for the first time: what is a tuple pattern? Maybe you could take this opportunity to add docstrings / comments to `handle_destruct_fields` and `handle_destruct` :)

Reply:
docstrings added